pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,963 @@
+		/// If start execution block number is not specified, then the schedule is planned in the consequent block.
+		///
+		/// The order will be executed within the configured AMM trade pool
+		///
+		/// Parameters:
+		/// - `origin`: schedule owner
+		/// - `schedule`: schedule details
+		/// - `start_execution_block`: start execution block for the schedule
+		///
+		/// Emits `Scheduled` and `ExecutionPlanned` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::schedule())]
+		#[transactional]
+		pub fn schedule(
+			origin: OriginFor<T>,
+			schedule: Schedule<T::Asset, BlockNumberFor<T>>,
+			start_execution_block: Option<BlockNumberFor<T>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			Self::ensure_that_next_blocknumber_bigger_than_current_block(start_execution_block)?;
+			Self::ensure_that_total_amount_is_bigger_than_storage_bond(&schedule)?;
+			Self::ensure_that_sell_amount_is_bigger_than_transaction_fee(&schedule)?;
+
+			let next_schedule_id = Self::get_next_schedule_id()?;
+
+			Schedules::<T>::insert(next_schedule_id, &schedule);
+			ScheduleOwnership::<T>::insert(next_schedule_id, who.clone());
+
+			let blocknumber_for_first_schedule_execution =
+				start_execution_block.unwrap_or_else(|| Self::get_next_block_mumber());
+			Self::plan_schedule_for_block(blocknumber_for_first_schedule_execution, next_schedule_id)?;
+
+			let currency_for_reserve = match schedule.order {
+				Order::Buy { asset_in, .. } => asset_in,
+				Order::Sell { asset_in, .. } => asset_in,
+			};
+
+			ensure!(
+				T::Currency::can_reserve(
+					currency_for_reserve.into(),
+					&who,
+					schedule.total_amount.into()
+				),
+				Error::<T>::InsufficientBalanceForTotalAmount
+			);
+
+			T::Currency::reserve_named(
+				&reserve_identifier(next_schedule_id),
+				currency_for_reserve.into(),
+				&who,
+				schedule.total_amount.into(),
+			)?;
+
+			Self::deposit_event(Event::Scheduled {
+				id: next_schedule_id,
+				who: who.clone(),
+			});
+			Self::deposit_event(Event::ExecutionPlanned {
+				id: next_schedule_id,
+				who,
+				block: blocknumber_for_first_schedule_execution,
+			});
+
+			Ok(())
+		}
+
+		/// Pause the DCA schedule planned in the given block number
+		///
+		/// Parameters:
+		/// - `origin`: schedule owner
+		/// - `schedule_id`: schedule id
+		/// - `next_execution_block`: block number where the DCA is planned to be executed
+		///
+		/// Emits `Paused` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::pause())]
+		#[transactional]
+		pub fn pause(
+			origin: OriginFor<T>,
+			schedule_id: ScheduleId,
+			next_execution_block: BlockNumberFor<T>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			Self::ensure_that_origin_is_schedule_owner(schedule_id, &who)?;
+
+			Self::remove_schedule_id_from_next_execution_block(schedule_id, next_execution_block)?;
+			Suspended::<T>::insert(schedule_id, ());
```

Comment:
what's the difference between suspended and paused? if they are the same, use the same terminology to avoid confusion

Reply:
suspended means that it is paused by the user, or a dca execution fails due to some unexpected or trade limit error

terminated means that the complete DCA is finished or cancelled by the user.

documented in the suspended event declaration, as requested ;)

Reply:
pause happens by the user, and it will make the order suspended
if an order fails for some reason, it will be also suspended.